src/App/Serialization/JsonRpcResponseErrorNormalizer.php

```diff
@@ -0,0 +1,151 @@
+    {
+        return $this->composeDebugErrorData($error->getPrevious() ?? $error);
+    }
+
+    /**
+     * @param \Throwable $error
+     * @return array
+     */
+    private function composeDebugErrorData(\Throwable $error) : array
+    {
+        $data = [
+            '_class' => get_class($error),
+            '_code' => $error->getCode(),
+            '_message' => $error->getMessage(),
+        ];
+
+        $trace = $this->filterErrorTrace($error->getTrace());
+        if (!empty($trace)) {
+            $data['_trace'] = $trace;
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param array $trace raw exception stack trace.
+     * @return array simplified stack trace.
+     */
+    private function filterErrorTrace(array $trace): array
+    {
+        $trace = array_slice($trace, 0, $this->maxTraceSize);
+
+        $result = [];
+        foreach ($trace as $entry) {
+            if (array_key_exists('args', $entry)) {
+                if ($this->showTraceArguments) {
+                    if ($this->simplifyTraceArguments) {
+                        $entry['args'] = $this->simplifyArguments($entry['args']);
+                    }
+                } else {
+                    unset($entry['args']);
+                }
+            }
+
+            $result[] = $entry;
+        }
+
+        return $result;
+    }
+
+    /**
+     * Converts arguments array to their simplified representation.
+     *
+     * @param array $args arguments array to be converted.
+     * @return string string representation of the arguments array.
+     */
+    private function simplifyArguments(array $args) : string
+    {
+        $count = 0;
+
+        $isAssoc = $args !== array_values($args);
+
+        foreach ($args as $key => $value) {
+            $count++;
+
+            if ($count >= 5) {
+                if ($count > 5) {
+                    unset($args[$key]);
+                } else {
+                    $args[$key] = '...';
+                }
+
+                continue;
+            }
+
+            if (is_object($value)) {
+                $args[$key] = get_class($value);
+            } elseif (is_bool($value)) {
+                $args[$key] = $value ? 'true' : 'false';
+            } elseif (is_string($value)) {
+                if (strlen($value) > 64) {
+                    $args[$key] = "'" . substr($value, 0, 64) . "...'";
+                } else {
+                    $args[$key] = "'" . $value . "'";
+                }
+            } elseif (is_array($value)) {
+                $args[$key] = '[' . $this->simplifyArguments($value) . ']';
+            } elseif ($value===null) {
+                $args[$key] = 'null';
+            } elseif (is_resource($value)) {
+                $args[$key] = 'resource';
+            }
+
+            if (is_string($key)) {
+                $args[$key] = "'" . $key . "' => " . $args[$key];
+            } elseif ($isAssoc) {
```

Comment:
I don't understand that condition. Shouldn't be just a `else` ?

Reply:
Indeed.
Fixes.

Reply:
Although, no - original condition is correct.
Reverted.

In this line we check if array contains both numeric keys and string keys or has keys reordered. If it does - we display the keys.
If array contains only numeric keys in assoc order we do not display keys at all, creating much cleaner output.

Reply:
For list-type array:

```php
[
    'foo',
    'bar',
]
```

representation will be:

```
['foo', 'bar']
```

But for the following array:

```php
[
    9 => 'nine',
    5 => 'five'
]
```

representation will be:

```
[9 => 'nine', 5 => 'five']
```

For the array with mixed keys:

```php
[
    'foo',
    'name' => 'bar',
]
```

representation will be:

```
[0 => 'foo', 'name' => 'bar']
```

Reply:
Got it 👍 Test params updated accordingly